runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -293,11 +293,116 @@ initImpl(J9VMThread *currentThread, j9object_t membernameObject, j9object_t refO
 	}
 }
 
+typedef struct LocalJ9UTF8Buffer {
+	LocalJ9UTF8Buffer(char *buffer, size_t length)
+	{
+		utf8 = reinterpret_cast<J9UTF8 *>(buffer);
+		cursor = J9UTF8_DATA(buffer);
+		capacity = length - sizeof(J9UTF8::length);
+	}
+
+	/**
+	 * Calculate the remaining slots in the buffer.
+	 * @return number of remaining slots in the buffer
+	 */
+	size_t remaining()
+	{
+		size_t cursorOffset = static_cast<size_t>(cursor - J9UTF8_DATA(utf8));
+		Assert_JCL_true(cursorOffset <= capacity);
+		return capacity - cursorOffset;
+	}
+
+	/**
+	 * Put a character into the buffer at the cursor, then advance the cursor.
+	 * @param[in] c The character to put into the buffer
+	 */
+	void putCharAtCursor(const char c)
+	{
+		size_t cursorOffset = static_cast<size_t>(cursor - J9UTF8_DATA(utf8));
+		Assert_JCL_true(cursorOffset < capacity);
+		*cursor = c;
+		cursor += 1;
+	}
+
+	/**
+	 * Put a null terminator into the buffer at the cursor, no advancing.
+	 */
+	void putNullTerminator()
+	{
+		size_t cursorOffset = static_cast<size_t>(cursor - J9UTF8_DATA(utf8));
+		Assert_JCL_true(cursorOffset <= capacity);
+		*cursor = '\0';
+	}
+
+	/**
+	 * Advance the cursor N slots.
+	 * @param[in] N The number of slots to advance the cursor by
+	 */
+	void advanceN(size_t n)
+	{
+		size_t cursorOffset = static_cast<size_t>(cursor - J9UTF8_DATA(utf8));
+		Assert_JCL_true(cursorOffset + n <= capacity);
+		cursor += n;
+	}
+
+	/**
+	 * Null-terminates the data, and sets the J9UTF8 length from a cursor-position calcuation.
+	 */
+	void commitLength()
+	{
+		putNullTerminator();
+		J9UTF8_SET_LENGTH(utf8, static_cast<U_16>(cursor - J9UTF8_DATA(utf8)));
+	}
+
+	J9UTF8 *utf8; /**< Pointer to the J9UTF8 struct */
+	size_t capacity; /**< Capacity of the J9UTF8 data buffer */
+	U_8 *cursor; /**< Pointer to current position in J9UTF8 data buffer */
+} LocalJ9UTF8Buffer;
+
+/**
+ * Adds a character corresponding to a primitive type into a signature buffer.
+ * @param[in] vm J9JavaVM instance for current thread
+ * @param[in] clazz The class of signature character of interest
+ * @param[in,out] stringBuffer The signature buffer to place the character into
+ */
+static void
+sigForPrimitiveOrVoidInout(J9JavaVM *vm, J9Class *clazz, LocalJ9UTF8Buffer *stringBuffer)
```

Comment:
I suggest just directly returning the character and do the putChar operation in caller, it makes the code a lot easier to read.

Reply:
This also helps to have it reused in the allocation path instead of needing two versions.